ebpfsvc/rpc_interface.idl

```diff
@@ -13,59 +13,61 @@ import "wtypes.idl";
-    typedef uint64_t ebpf_instruction;
+    interface ebpf_service_interface {
 
-    typedef [system_handle(sh_file)] HANDLE file_handle_t;
+        // TODO: This is a duplicate of EbpfMapDescriptor structure
+        // defined in spec_type_descriptor.hpp. See if that file
+        // can be included so that we do not have to redefine this.
+        typedef struct _ebpf_map_descriptor
+        {
+            int original_fd;
+            uint32_t type; // Platform-specific type value in ELF file.
+            unsigned int key_size;
+            unsigned int value_size;
+            unsigned int max_entries;
+            unsigned int inner_map_fd;
+        } ebpf_map_descriptor;
 
-    typedef struct _fd_handle_map
-    {
-        uint32_t file_descriptor;
-        file_handle_t handle;
-    } fd_handle_map;
+        typedef[system_handle(sh_file)] HANDLE file_handle_t;
 
-    typedef struct _ebpf_program_load_info
-    {
-        // Optional file name with full path.
-        [string] char* file_name;
-        // Optional program name.
-        [string] char* program_name;
-        GUID program_type;
-        file_handle_t program_handle;
-        ebpf_execution_context_t execution_context;
-        uint32_t map_count;
-        [size_is(map_count)] fd_handle_map* handle_map;
-        uint32_t instruction_count;
-        [size_is(instruction_count)] ebpf_instruction* instructions;
-    } ebpf_program_load_info;
+        typedef struct _fd_handle_map
+        {
+            uint32_t file_descriptor;
+            file_handle_t handle;
+        } fd_handle_map;
 
-    typedef struct _ebpf_program_verify_info
-    {
-        GUID program_type;
-        ebpf_execution_context_t execution_context;
-        uint32_t map_count;
-        [size_is(map_count)] fd_handle_map* handle_map;
-        uint32_t instruction_count;
-        [size_is(instruction_count)] ebpf_instruction* instructions;
-    } ebpf_program_verify_info;
+        typedef struct _ebpf_program_load_info
+        {
+            // Optional file name with full path.
+            [string] char* file_name;
+            // Optional program name.
+            [string] char* program_name;
+            GUID program_type;
+            file_handle_t program_handle;
+            ebpf_execution_context_t execution_context;
+            uint32_t map_count;
+            [size_is(map_count)] fd_handle_map* handle_map;
+            uint32_t byte_code_size;
+            [size_is(byte_code_size)] uint8_t* byte_code;
+        } ebpf_program_load_info;
 
-    ebpf_result_t
-    ebpf_verify_and_jit_program(
-        [in] ebpf_program_load_info* info,
-        [out] uint32_t* logs_size,
-        [out, size_is(,*logs_size)]char** logs);
+        typedef struct _ebpf_program_verify_info
+        {
+            GUID program_type;
+            ebpf_execution_context_t execution_context;
+            uint32_t map_count;
+            [size_is(map_count)] ebpf_map_descriptor* map_descriptors;
```

Comment:
You can't trust the value you get from the client. You need to query kernel using the handle to get the actual value.

Reply:
Just realized this is the verify only path, so no one is trusting the result of the verification, so spoofed map_descriptors is a non-issue.

Reply:
Yes. Also since this is verification only, there will be no maps actually created. So EC will have no knowledge of these maps.